pkg/deploy/controller/deployment_config_controller.go

```diff
@@ -20,73 +26,185 @@ import (
+
+	deploymentStore := cache.NewStore(cache.MetaNamespaceKeyFunc)
+	cache.NewReflector(deploymentLW, &kapi.ReplicationController{}, deploymentStore).RunUntil(stop)
+
+	// Create a deploymentClient impl backed by the kube client
+	deploymentClient := &deploymentConfigControllerClientImpl{
+		CreateDeploymentFunc: func(namespace string, deployment *kapi.ReplicationController) (*kapi.ReplicationController, error) {
+			return kClient.ReplicationControllers(namespace).Create(deployment)
+		},
+	}
+
+	// Sync every fullSyncPeriod, or every time a deploymentConfig watch event is observed.
+	// TODO: Compress events to prevent sync spamming.
+	// TODO: Extract this into something more generic.
+	sync := make(chan struct{})
+	go func() {
+		// TODO: error handling
+		configWatch, _ := osClient.DeploymentConfigs(kapi.NamespaceAll).Watch(labels.Everything(), labels.Everything(), "")
+		ticker := time.NewTicker(fullSyncPeriod)
+		for {
+			select {
+			case <-ticker.C:
+				sync <- struct{}{}
+			case event, ok := <-configWatch.ResultChan():
+				if !ok || event.Type == watch.Error {
+					glog.Errorf("Re-establishing deploymentConfig watch")
+					configWatch, _ = osClient.DeploymentConfigs(kapi.NamespaceAll).Watch(labels.Everything(), labels.Everything(), "")
+					continue
+				}
+				sync <- struct{}{}
+			case <-stop:
+				return
+			}
+		}
+	}()
+
+	return &DeploymentConfigController{
+		client:                deploymentClient,
+		deploymentConfigStore: deploymentConfigStore,
+		deploymentStore:       deploymentStore,
+		codec:                 codec,
+		sync:                  sync,
+		stop:                  stop,
+		handleError: func(err error) {
+			util.HandleError(err)
+		},
+	}
 }
 
-// Process DeploymentConfig events one at a time.
+// Run performs a full sync every period, or when a message is received on sync,
+// until a message is received on the stop channel.
 func (c *DeploymentConfigController) Run() {
-	go util.Until(c.HandleDeploymentConfig, 0, c.Stop)
+	go util.Until(func() {
+		select {
+		case <-c.sync:
+			c.syncAll()
+		}
+	}, 0, c.stop)
 }
 
-// Process a single DeploymentConfig event.
-func (c *DeploymentConfigController) HandleDeploymentConfig() {
-	config := c.NextDeploymentConfig()
-	deploy, err := c.shouldDeploy(config)
-	if err != nil {
-		util.HandleError(fmt.Errorf("unable to decide whether to redeploy %s: %v", labelFor(config), err))
-		return
+// syncAll performs a sync on every config in the store.
+func (c *DeploymentConfigController) syncAll() {
+	for _, config := range c.deploymentConfigStore.List() {
+		c.syncOne(config.(*deployapi.DeploymentConfig))
 	}
-	if !deploy {
+}
+
+// sync examines the current state of a DeploymentConfig, and creates a new
+// deployment for the config if the following conditions are true:
+//
+//   1. The config version is greater than 0
+//   2. No deployment exists corresponding to  the config's version
+//
+// Processing errors are delegated to handleError, and the config will likely
```

Comment:
I don't like this pattern - this function should be returning an error, and the higher level logic is what processes that error.